### PR TITLE
Remove unused imports

### PR DESCRIPTION
**Pull request by flying-sheep**
Clarify API packages, remove unused imports

### CODE REVIEW

File: anndata/_io/zarr.py [hunk 1]
```diff
@@ -136,8 +145,6 @@ def read_dataframe_legacy(dataset: zarr.Array) -> pd.DataFrame:
 
 @report_read_key_on_error
 def read_dataframe(group) -> pd.DataFrame:
-    from .specs import _REGISTRY
```
Comment (flying-sheep) [inline anndata/_io/zarr.py:139]: if this is necessary, it should have a comment explaining why

File: anndata/_io/h5ad.py [hunk 2]
```diff
@@ -41,8 +51,6 @@ def write_h5ad(
     dataset_kwargs: Mapping = MappingProxyType({}),
     **kwargs,
 ) -> None:
-    from anndata.experimental import write_dispatched
```
Comment (flying-sheep) [inline anndata/_io/h5ad.py:44]: This is less likely to be a import with side effects, but it could be too: is it?

Comment (flying-sheep) [reply anndata/_io/h5ad.py:44]: also not necessary

File: anndata/__init__.py [hunk 3]
```diff
@@ -34,4 +34,24 @@
...
 
 # backwards compat / shortcut for default format
-from ._io import read_h5ad as read
+from ._io import read_h5ad as read  # noqa: F401
+
+__all__ = [
+    "__version__",
+    "AnnData",
+    "concat",
+    "Raw",
```
Comment (ivirshup) [inline anndata/__init__.py:43]:
```suggestion
    "Raw",
    "experimental",
```

Right?

File: pyproject.toml [hunk 4]
```diff
@@ -133,6 +131,8 @@ select = [
     "F",
     "W",
 ]
+# Disable autofix for unused imports (`F401`).
+unfixable = ["F401"]
```
Comment (ivirshup) [inline pyproject.toml:135]: Why?

File: anndata/_io/h5ad.py [hunk 5]
```diff
@@ -30,6 +30,16 @@
...
+
+__all__ = [
+    "read_elem",
+    "write_elem",
+    "read_h5ad",
+    "write_h5ad",
+    "read_dataset",
+]
+
+
```
Comment (ivirshup) [inline anndata/_io/h5ad.py:42]: Why put this in a private module?

File: anndata/__init__.py (hunk 3 above)
Comment (flying-sheep) [reply anndata/__init__.py:43]: No, if you have a submodule that nothing has imported yet, the submodule is not a member of the super module:

```console
$ mkdir mod
$ touch mod/submod.py
$ touch mod/__init__.py
$ python -c 'import mod; mod.submod'
AttributeError: Module `mod` has no member `submod`
$ python -c 'import mod.submod'  # works
```

File: pyproject.toml (hunk 4 above)
Comment (flying-sheep) [reply pyproject.toml:135]: If you have your editor set up to do them for you, it can be annoying to have it remove imports that are already written but not used yet.

I’m OK with both, I’ve changed my habits to never do the above, but I thought it might be annoying for others. Your call!

File: anndata/_io/h5ad.py (hunk 5 above)
Comment (flying-sheep) [reply anndata/_io/h5ad.py:42]: It clarifies what’s imported from elsewhere. That’s the list of functions that `anndata._io.zarr` has in an equivalent way. But we can get rid of it if you prefer.

File: pyproject.toml (hunk 4 above)
Comment (ivirshup) [reply pyproject.toml:135]: I would really prefer that it makes the change on pre-commit

File: anndata/_io/h5ad.py (hunk 5 above)
Comment (flying-sheep) [reply anndata/_io/h5ad.py]: he prefers

File: anndata/_io/specs/__init__.py [hunk 6]
```diff
@@ -1,2 +1,13 @@
...
+
+__all__ = [
+    "methods",
+    "write_elem",
+    "get_spec",
+    "read_elem",
+    "Reader",
+    "Writer",
+    "IOSpec",
+]
```
Comment (ivirshup) [inline anndata/_io/specs/__init__.py:13]: Again, private module defining exports

```suggestion
```

File: anndata/_io/read.py [hunk 7]
```diff
@@ -27,6 +27,19 @@ def read_zarr(*_, **__):
...
+    "read_csv",
+    "read_excel",
+    "read_umi_tools",
+    "read_hdf",
+    "read_loom",
+    "read_mtx",
+    "read_text",
+]
+
+
```
Comment (ivirshup) [inline anndata/_io/read.py:42]:
```suggestion
```

File: anndata/_io/__init__.py [hunk 8]
```diff
@@ -9,5 +9,22 @@
...
+    "read_loom",
+    "read_mtx",
+    "read_text",
+    "read_zarr",
+    "read_h5ad",
+    "write_csvs",
+    "write_loom",
+    "write_zarr",
+    "h5ad",
+]
```
Comment (ivirshup) [inline anndata/_io/__init__.py:30]: So, two things here:

* I don't think we should define namespaces for private modules
* I am up for a public `io` module, but I think we should do that in a separate PR

```suggestion
```

**Review by ivirshup — CHANGES_REQUESTED** [review]

File: anndata/_io/specs/__init__.py (hunk 6 above)
Comment (flying-sheep) [reply anndata/_io/specs/__init__.py:13]: please look at the file: we’re using it as an internal API module.

are you saying that we should get rid of all of those and directly import everything from where it’s defined?

**Review by ivirshup — APPROVED** [review]
